app/src/main/java/com/hfad/sapper/adapter/CellAdapter.kt

```diff
@@ -0,0 +1,115 @@
+package com.hfad.sapper.adapter
+
+import android.graphics.Color
+import android.util.Log
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.Button
+import androidx.recyclerview.widget.RecyclerView
+import com.hfad.sapper.R
+import com.hfad.sapper.data.Cell
+import com.hfad.sapper.ui.main.CLEAR_CELL
+import com.hfad.sapper.ui.main.FIELD_SIZE
+import com.hfad.sapper.ui.main.MINE_CELL
+import com.hfad.sapper.ui.main.TRIGGER_MINE_CELL
+
+class CellAdapter(private var playingField: MutableList<Cell> = mutableListOf()) :
+    RecyclerView.Adapter<CellAdapter.ViewHolder>() {
+    private var itemListener: ItemClickListener? = null
+
+    fun setListener(itemClickListener: ItemClickListener) {
+        this.itemListener = itemClickListener
+    }
+
+    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(viewGroup.context)
+            .inflate(R.layout.item_cell, viewGroup, false)
+
+        return ViewHolder(view)
+    }
+
+    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
+        val cell = playingField[position]
+        when (cell.valueCell) {
+            CLEAR_CELL, MINE_CELL -> {
+                viewHolder.cellButton.setBackgroundColor(Color.GRAY)
+            }
+            TRIGGER_MINE_CELL -> {
+                viewHolder.cellButton.setBackgroundColor(Color.RED)
+            }
+            else -> {
+                viewHolder.cellButton.setBackgroundColor(Color.WHITE)
+                viewHolder.cellButton.text = countMineAround(position).toString()
+            }
+        }
+        viewHolder.cellButton.setOnClickListener {
+            itemListener?.onItemClick(playingField, position)
+        }
+    }
+
+    override fun getItemCount() = playingField.size
+
+    fun setPlayingField(playingField: MutableList<Cell>) {
+        this.playingField = playingField
+        notifyDataSetChanged()
+    }
+
+    fun updatePlayingField(position: Int, value: Int) {
+        playingField[position].valueCell = value
+        notifyItemChanged(position)
+    }
+
+    private fun countMineAround(position: Int): Int {
```

Comment:
Nice!

Reply:
What is not nice in this function, is that you recalculate the value each time `onBindViewHolder` triggered. Such an approach could freeze your screen during scroll. Here it's not a problem. But approach - is a way to a bad performance. 
Also, it's not a responsibility of the Adapter to calculate such value. But we'll discuss this moment during our next lesson.